// components/CTracksComp/utils/tracks.js
function horizontalPointOptions (overrides) {
  return Object.assign({
    labelColor: "black",
    labelPosition: "topLeft",
    axisPositionHorizontal: "right",
    pointColor: "red",
    pointSize: 2,
    valueScaling: "linear",
    trackBorderWidth: 1,
    trackBorderColor: "grey",
    labelTextOpacity: 0.4,
    name: "template_label"
  }, overrides)
}

function track (overrides, options) {
  return Object.assign(
    {
      name: "",
      server: "", // FUTURE CONFIG
      tilesetUid: "",
      uid: "",
      type: "horizontal-point",
      width: 770,
      height: 38,
      position: "top"
    },
    overrides,
    {
      options: {...options}
    }
  )
}

const chromosomeAxisTrack = {
  name: "Chromosome Axis",
  server: "http://higlass.io/api/v1", // FUTURE CONFIG
  tilesetUid: "Ajn_ttUUQbqgtOD4nOt-IA", // FUTURE CONFIG
  uid: "bpRZDog1QQuZ7DmLfsKwXw",
  type: "horizontal-chromosome-labels",
  options: {},
  width: 770,
  height: 30,
  position: "top"
}

const geneAnnotationsHG19Track = {
  name: "Gene Annotations (hg19)",
  server: "http://higlass.io/api/v1", // FUTURE CONFIG
  tilesetUid: "OHJakQICQD6gTD7skx4EWA", // FUTURE CONFIG
  uid: "NaYcUhtrQyeQhqz9gJOjxQ",
  type: "horizontal-gene-annotations",
  options: {
    labelColor: "black",
    labelPosition: "hidden",
    plusStrandColor: "blue",
    minusStrandColor: "red",
    trackBorderWidth: 0,
    trackBorderColor: "black",
    name: "Gene Annotations (hg19)"
  },
  width: 20,
  height: 55,
  position: "top"
}

export default {
  horizontalPointOptions,
  track,
  chromosomeAxisTrack,
  geneAnnotationsHG19Track
}
