// components/CTracksComp/utils/tracks.js
function horizontalOptions (overrides) {
  return Object.assign({
    labelColor: "black",
    labelPosition: "topLeft",
    axisPositionHorizontal: "right",
    valueScaling: "linear",
    trackBorderWidth: 1,
    trackBorderColor: "grey",
    labelTextOpacity: 0.4
  }, overrides)
}

function horizontalPointOptions (overrides, primaryColor) {
  return Object.assign(
    horizontalOptions(),
    {
      pointColor: primaryColor || "red"
    },
    overrides
  )
}

function horizontalLineOptions (overrides, primaryColor) {
  return Object.assign(
    horizontalOptions(),
    {
      lineStrokeColor: primaryColor || "red"
    },
    overrides
  )
}

function horizontalBarOptions (overrides, primaryColor) {
  return Object.assign(
    horizontalOptions(),
    {
      barFillColor: primaryColor || "red"
    },
    overrides
  )
}

function horizontalRangeOptions (overrides, primaryColor) {
  return Object.assign(
    horizontalOptions(),
    {
      minMaxColor: primaryColor || "red",
      stdFillColor: primaryColor || "red",
      stdStrokeColor: primaryColor || "red",
      vLineColor: "grey"
    },
    overrides
  )
}

function getPrimaryColor (track) {
  switch (track.type) {
    case 'horizontal-point':
      return track.options.pointColor
    case 'horizontal-line':
      return track.options.lineStrokeColor
    case 'horizontal-bar':
      return track.options.barFillColor
    case 'range':
      return track.options.mode === 'whisker' ? track.options.stdFillColor : track.options.minMaxColor
    default:
      return 'red'
  }
}

function track (overrides, options) {
  return Object.assign(
    {
      name: "",
      server: "", // FUTURE CONFIG
      tilesetUid: "",
      uid: "",
      type: "horizontal-point",
      datatype: "vector",
      height: 38
    },
    overrides,
    {
      options: {
        showTooltip: true,
        ...options
      }
    }
  )
}

const chromosomeAxisTrack = {
  name: "Chromosome Axis",
  server: "http://higlass.io/api/v1", // FUTURE CONFIG
  tilesetUid: "Ajn_ttUUQbqgtOD4nOt-IA", // FUTURE CONFIG
  uid: "bpRZDog1QQuZ7DmLfsKwXw",
  type: "horizontal-chromosome-labels",
  options: {},
  height: 30,
  position: "top"
}

const geneAnnotationsHG19Track = {
  name: "Gene Annotations (hg19)",
  server: "http://higlass.io/api/v1", // FUTURE CONFIG
  tilesetUid: "OHJakQICQD6gTD7skx4EWA", // FUTURE CONFIG
  uid: "NaYcUhtrQyeQhqz9gJOjxQ",
  type: "horizontal-gene-annotations",
  options: {
    labelColor: "black",
    labelPosition: "hidden",
    plusStrandColor: "blue",
    minusStrandColor: "red",
    trackBorderWidth: 0,
    trackBorderColor: "black",
    name: "Gene Annotations (hg19)"
  },
  height: 55,
  position: "top"
}

const trackDataTypes = {
  'bedlike': ['bedlike'],
  'gene-annotations': ['horizontal-gene-annotations'],
  'matrix': ['horizontal-heatmap'], // unsupported 2d options: 'heatmap'
  '2d-rectangle-domains': ['horizontal-2d-rectangle-domains'], // unsupported 2d options: '2d-rectangle-domains'
  'vector': ['horizontal-point', 'horizontal-bar', 'horizontal-line', 'horizontal-1d-heatmap'], // unsupported 2d options: 'vertical-1d-heatmap'
  'multivec': ['horizontal-stacked-bar', 'basic-multiple-line-chart', 'basic-multiple-bar-chart'],
  'chromsizes': ['horizontal-chromosome-labels'], // unsupported 2d options: '2d-chromosome-grid'
  'annotations': ['horizontal-1d-annotations'] // unsupported 2d options: 'vertical-1d-annotations'
}

const fileTypes = [
  'hitile',
  'bigwig',
  'beddb',
  'chromsizes-tsv',
  'multivec'
]

export default {
  options: {
    'horizontal-point': horizontalPointOptions,
    'horizontal-line': horizontalLineOptions,
    'horizontal-bar': horizontalBarOptions,
    'range': horizontalRangeOptions
  },
  getPrimaryColor,
  track,
  chromosomeAxisTrack,
  geneAnnotationsHG19Track,
  trackDataTypes,
  fileTypes
}
