// components/CTracksComp/CTracksComp.js
import React from 'react'
import hamradio from 'hamradio'
import axios from 'axios'

import HiglassUI from './HiglassUI'
import CNVTable from './CNVTable'
import ChromView from './ChromView'
import TracksMenu from './TracksMenu'
import ServerAdd from './ServerAdd'
import RegisterURL from './RegisterURL'


import './CTracksComp.css'


class CTracksComp extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      expanded: true
    }
    this.subscriptions = this.makeSubscriptions()
  }

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions() {
    return [
      hamradio.subscribe(
        'registerURL',
        (name, {server, ...data}) => {
          axios.post(`${server}/register_url/`, data).then(response => {
            hamradio.publish('server/updated', server)
          })
        }
      )
    ]
  }

  hideTools = () => {
    this.setState({expanded: false})
  }

  showTools = () => {
    this.setState({expanded: true})
  }

  render() {
    return (
      <div className = "TopContainer">

        {!this.state.expanded || <div className="LeftPanel">
          <div className="Box PanelBox">
            <ChromView chromInfo = {this.props.chromInfo} />
          </div>
          <div className="Box PanelBox">
            <ServerAdd/>
          </div>
          <div className="Box PanelBox">
            <RegisterURL/>
          </div>
          <div className="Box PanelBox">
            <TracksMenu/>
          </div>
        </div>}

        <div className = "Expander">
          {this.state.expanded
            ? <div className = "ExpandButton" onClick={this.hideTools}> &lt; </div>
            : <div className = "ExpandButton" onClick={this.showTools}> &gt; </div>
          }
        </div>

        <div className = "RightPanel">
          <div className = "Box">
            <HiglassUI
              uid = {this.props.uid}
              chromInfo = {this.props.chromInfo}
            />
          </div>
          <CNVTable />
        </div>

      </div>
    )
  }
}

export default CTracksComp;
