// components/CTracksComp/CTracksComp.js
import React from 'react';

import HiglassUI from './HiglassUI';
import CNVTable from './CNVTable';
import ChromView from './ChromView';
import TracksMenu from './TracksMenu';
import {testCNVTable} from './TestCNVTable.js';

import ReactFileReader from 'react-file-reader';
import Papa from 'papaparse';

import hamradio from 'hamradio'

//import '../node_modules/bootstrap/dist/css/bootstrap.min.css';
import './CTracksComp.css';

//import './Components/HiGlassLauncher.scss';


// Parse CNV BED file as a CSV file
function ParseFile(file, callback) {
  Papa.parse(file, {
    //download: true,
    newline: "\n",
    header: true,
    dynamicTyping: true,
    error: function(error) {
      console.log("ERROR Parse:",error);
    },
    complete: function(results) {
      callback(results.data);
    }
  });
}


class CTracksComp extends React.Component {
  constructor (props) {
    super(props);

    // CNVData: BED File information
    // CNVSelection: selected CNV samples from analysis
    this.state = {
      CNVData: null,
      CNVSelection: null
    };

    this.subscriptions = this.makeSubscriptions()

    //Binding functions
    this.UpdateCNVData = this.UpdateCNVData.bind(this);
    this.ProcessCNVFile = this.ProcessCNVFile.bind(this);
  }

  componentWillMount() {

  }

  componentDidMount() {
    this.setState({
      CNVData: testCNVTable,
    });
  }

  componentDidUpdate() {

  }

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions() {
    return []
  }

  // Process CNV BED file: upload, parse and update state
  ProcessCNVFile (files) {
    // console.log("Processing CNVFile...");
    // Loading file
    var reader = new FileReader();
    reader.onload = function(e) {
      // Use reader.result
      // var text = reader.result;
    }
    //reader.readAsText(files[0]);
    this.UpdateCNVSelection(null);
    ParseFile(files[0], this.UpdateCNVData);
  }

  // Update state of CNV data
  UpdateCNVData(data) {
    //console.log("CNVData:", JSON.stringify(data, null, 2));
    this.setState(function () {
      return {
        CNVData: data
      }
    });
  }

  handleCNVSelection (CNVSelection) {
    //console.log("Table Selection: ", CNVSelection);
    this.UpdateCNVSelection(CNVSelection);
  }

  // Update state of CNV selection
  UpdateCNVSelection(selection) {
    this.setState(function () {
      return {
        CNVSelection: selection
      }
    });
  }

  render() {
    return (
      <div className = "TopContainer">

        <div className = "LeftPanel">
          <div className = "Box">
            <label>Loading input files</label>

            <div className = "FileReader">
              <ReactFileReader handleFiles={this.ProcessCNVFile} fileTypes={'.bed, .tsv'}>
                <button className="btn btn-default">Load CNV BED File</button>
              </ReactFileReader>
            </div>

          </div>

          <div className = "Box">
            <label>Higlass - Update bottom view</label>
            <ChromView chromInfo = {this.props.chromInfo} />
          </div>

          <div className = "Box">
            <TracksMenu />
          </div>

        </div>

        <div className = "RightPanel">
          <div className = "RightBox">
            <HiglassUI
              uid = {this.props.uid}
              chromInfo = {this.props.chromInfo}
            />
          </div>
          <div className = "RightBox">
            {this.state.CNVData &&
            <CNVTable
              CNVData={this.state.CNVData}
              onCNVSelection={this.handleCNVSelection.bind(this)}
            />}
          </div>
        </div>

      </div>
    )
  }
}

export default CTracksComp;
