// components/CTracksComp/HiglassUI.js
import React from 'react';
import PropTypes from 'prop-types';

import * as hglib from 'higlass';
import hamradio from 'hamradio'
import './HiglassUI.css'

import 'higlass/build/hglib.css';

class HiglassUI extends React.Component {
  constructor(props) {
    super(props);

    this.publishLocation = this.publishLocation.bind(this);

    this.prevViewConfig = JSON.stringify(props.ViewConfig)
  }

  shouldComponentUpdate(newProps, newState) {
    if (this.prevViewConfig === newProps.ViewConfig) {
      return false;
    }

    //console.log('yes', newProps.ViewConfig);
    //console.log("shouldComponentUpdate: HiglassUI");
    this.prevViewConfig = newProps.ViewConfig;
    return true;
  }

  publishLocation (location) {
    if (this.props.chromInfo) {
      let chrome = location.xDomain
        .map(coord => this.props.chromInfo.absToChr(coord))
      hamradio.publish('viewport', chrome)
    } else {
      setTimeout(this.publishLocation, 10, location)
    }
  }

  launchHgLib(ViewConfig, options) {
    return (element) => {
      if (element && ViewConfig) {
        let api = hglib.viewer(
          element,
          ViewConfig,
          options
        );

        api.on('location', this.publishLocation, ViewConfig.views[1].uid)
        window.hgApi = api
      }
    }
  }


  componentDidMount() {
    this.props.onHiglassUpdated();
  }

  componentDidUpdate() {
    this.props.onHiglassUpdated();
  }

  render () {
  	return (
        <div className = "higlass"
          ref={this.launchHgLib(
            JSON.parse(this.props.ViewConfig),
            { bounded: false }
          )}>
        </div>
		)
  }
}

HiglassUI.propTypes = {
  ViewConfig: PropTypes.string,
  onHiglassUpdated: PropTypes.func,
}

export default HiglassUI;
