// components/CTracksComp/RegisterURL.js
import React from 'react'
import hamradio from 'hamradio'
import axios from 'axios'

import './RegisterURL.css'

import tracks from './utils/tracks'
import TextInput from './basicComponents/TextInput'
import LabelLayout from './basicComponents/LabelLayout'
import Selector from './basicComponents/Selector'
import TextButton from './basicComponents/TextButton'

class RegisterURL extends React.Component {
  constructor(props) {
    super(props)

    this.subscriptions = this.makeSubscriptions()

    this.state = {
      url: '',
      name: '',
      server: this.props.servers.length ? this.serverElement(this.props.servers[0], 0) : {value: null, label:'', api: ''},
      datatype: {value: 'vector', label: 'vector'},
      filetype: {value: 'hitile', label: 'hitile'},
      error: null
    }
  }

  makeSubscriptions = () => {
    return [
      hamradio.subscribe(
        'url/register/error',
        (name, data) => {
          let serverindex = this.props.servers.findIndex(server => server.api === data.data.server)
          let server = serverindex === -1
            ? {value: null, label: data.data.server, api: data.data.server}
            : this.serverElement(this.props.servers[serverindex], serverindex)
          this.setState({
            url: data.data.url,
            name: data.data.name,
            server,
            datatype: data.data.datatype,
            filetype: data.data.filetype,
            error: data.error.response.data.error
          })
          setTimeout(this.setState.bind(this), 5000, {error: null})
        }
      )
    ]
  }

  serverElement = (server, order) => {
    return {
      label: server.name || server.api,
      value: order,
      api: server.api
    }
  }

  done = () => {
    this.setState({error: "Saving..."})
    console.log(this.state.server.value, this.state.url)
    if (this.state.server.value !== null && this.state.url !== '') {
      console.log('publishing url')
      hamradio.publish('url/register', {
        server: this.state.server.api,
        url: this.state.url,
        name: this.state.name,
        datatype: this.state.datatype.value,
        filetype: this.state.filetype.value,
        coordSystem: 'b37'
      }).then(() => {
        this.setState({
          url: '',
          name: '',
          server: this.props.servers.length ? this.serverElement(this.props.servers[0], 0) : {value: null, label:'', api: ''},
          datatype: {value: 'vector', label: 'vector'},
          filetype: {value: 'hitile', label: 'hitile'},
          error: null
        })
      }).catch(() => {
        this.setState({error: null})
      })
    }
  }

	render () {
		return (
      <div>
        <LabelLayout label="Register URL"
          inner=<TextInput value={this.state.url} update={val => this.setState({url:val})}/>
        />
        <LabelLayout label="Name"
          inner=<TextInput value={this.state.name} update={val => this.setState({name:val})}/>
        />
        <LabelLayout label="Registry Server"
          inner=<Selector
            value={this.state.server}
            options={this.props.servers.map((server, index) => this.serverElement(server, index))}
            onChange={val => this.setState({server: val})}
          />
        />
        <LabelLayout label="Data Type"
          inner=<Selector
            value={this.state.datatype}
            options={Object.keys(tracks.trackDataTypes).map(datatype => { return {label: datatype, value: datatype} })}
            onChange={val => this.setState({datatype: val})}
          />
        />
        <LabelLayout label="File Type"
          inner=<Selector
            value={this.state.filetype}
            options={tracks.fileTypes.map(filetype => { return {label: filetype, value: filetype} })}
            onChange={val => this.setState({filetype: val})}
          />
        />
        {this.state.error
          ? <label className="WarningLabel">Failed to register url: {this.state.error}</label>
          : <TextButton onClick={this.done} text="Register" />
        }
      </div>
    )
	}
}


export default RegisterURL;
