// components/CTracksComp/ResilientData.js
import _ from 'lodash'
import trackUtils from './utils/tracks'

let servers = []
let tracks = {}
let variants = []
let features = {}
let tools = {
  zoom: true,
  servers: true,
  urls: true,
  tracks: false
}
let prefix = ''

export default {
  initialize (newPrefix) {
    prefix = newPrefix
    servers = JSON.parse((sessionStorage.getItem(`${prefix}/servers`))) || []
    tracks = JSON.parse((sessionStorage.getItem(`${prefix}/tracks`))) || {}
    variants = JSON.parse((sessionStorage.getItem(`${prefix}/variants`))) || []
    features = JSON.parse((sessionStorage.getItem(`${prefix}/features`))) || {}
    tools = JSON.parse((sessionStorage.getItem(`${prefix}/tools`))) || {
      zoom: true,
      servers: true,
      urls: true,
      tracks: true
    }
  },
  servers: {
    get: () => _.cloneDeep(servers),
    add: (server) => {
      servers.push(server)
      sessionStorage.setItem(`${prefix}/servers`, JSON.stringify(servers))
      return _.cloneDeep(servers)
    }
  },
  tracks: {
    get: () => _.cloneDeep(tracks),
    remove: (row, uid) => {
      if (tracks[row]) {
        if (tracks[row].tilesets[uid]) {
          delete tracks[row].tilesets[uid]
        }
        if (Object.keys(tracks[row].tilesets).length === 0) {
          delete tracks[row]
        }
        sessionStorage.setItem(`${prefix}/tracks`, JSON.stringify(tracks))
      }
      return _.cloneDeep(tracks)
    },
    set: (row) => {
      if (!tracks[row.row]) {
        tracks[row.row] = row
      } else {
        trackUtils.merge(tracks[row.row], row)
      }

      if (tracks[row.row].global === null) tracks[row].global = true
      if (tracks[row.row].focus === null) tracks[row].focus = true

      sessionStorage.setItem(`${prefix}/tracks`, JSON.stringify(tracks))
      return _.cloneDeep(tracks)
    }
  },
  variants: {
    get: () => _.cloneDeep(variants),
    set: (newVariants) => {
      variants = _.cloneDeep(newVariants)
      sessionStorage.setItem(`${prefix}/variants`, JSON.stringify(variants))
      return _.cloneDeep(variants)
    }
  },
  features: {
    get: () => _.cloneDeep(features),
    set: (featureset, newFeatures) => {
      if (features[featureset]) {
        _.mergeWith(features[featureset], newFeatures, (o, s, key) => key === 'locations' ? s : undefined)
      } else {
        features[featureset] = newFeatures
      }
      console.log(features)
      sessionStorage.setItem(`${prefix}/features`, JSON.stringify(features))
      return _.cloneDeep(features)
    }
  },
  tools: {
    get: () => _.cloneDeep(tools),
    set: (updates) => {
      _.merge(tools, updates)
      sessionStorage.setItem(`${prefix}/tools`, JSON.stringify(tools))
      return _.cloneDeep(tools)
    }
  }
}
