// components/CTracksComp/ResilientData.js
import hamradio from 'hamradio'

let servers = []
let tracks = []
let variants = []
let tools = {
  zoom: true,
  servers: true,
  urls: true,
  tracks: true
}

let subscriptions = []

export default {
  initialize (prefix) {
    servers = JSON.parse((sessionStorage.getItem(`${prefix}/servers`))) || []
    tracks = JSON.parse((sessionStorage.getItem(`${prefix}/tracks`))) || []
    variants = JSON.parse((sessionStorage.getItem(`${prefix}/variants`))) || []
    tools = JSON.parse((sessionStorage.getItem(`${prefix}/tools`))) || {
      zoom: true,
      servers: true,
      urls: true,
      tracks: true
    }

    subscriptions.forEach(sub => sub.unsubscribe())
    subscriptions = [
      hamradio.subscribe('track/add', (name, data) => {
        tracks.push(data)
        sessionStorage.setItem(`${prefix}/tracks`, JSON.stringify(tracks))
      }),
      hamradio.subscribe('track/modify', (name, data) => {
        tracks = tracks.filter(track => data.track.uid !== track.track.uid)
        tracks.push(data)
        sessionStorage.setItem(`${prefix}/tracks`, JSON.stringify(tracks))
      }),
      hamradio.subscribe('track/remove', (name, data) => {
        tracks = tracks.filter(track => data !== track.track.uid)
        sessionStorage.setItem(`${prefix}/tracks`, JSON.stringify(tracks))
      }),
      hamradio.subscribe('server/add', (name, data) => {
        servers.push(data)
        sessionStorage.setItem(`${prefix}/servers`, JSON.stringify(servers))
      }),
      hamradio.subscribe('cTracks/variants', (name, data) => {
        variants = data
        sessionStorage.setItem(`${prefix}/variants`, JSON.stringify(data))
      }),
      hamradio.subscribe('tools', (name, data) => {
        tools = Object.assign({}, tools, data)
        sessionStorage.setItem(`${prefix}/tools`, JSON.stringify(data))
      })
    ]
  },
  servers () {
    return [...servers]
  },
  tracks () {
    return [...tracks]
  },
  variants () {
    return [...variants]
  },
  tools () {
    return {...tools}
  }
}
