// components/CTracksComp/ServerAdd.js
import React from 'react'
import axios from 'axios'

import './ServerAdd.css'

import hamradio from 'hamradio'

class ServerAdd extends React.Component {
  constructor (props) {
    super(props)

    this.subscriptions = this.makeSubscriptions()

    this.state = {
      servers: [],
      addName: '',
      addAPI: '',
      error: null
    }
  }

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions = () => {
    return [
      hamradio.subscribe(
        'server/add',
        (name, server) => {
          this.setState({
            servers: this.state.servers.concat([server])
          })
        }
      )
    ]
  }

  addServer = () => {
    axios.get(`${this.state.addAPI}/tilesets`)
      .then(response => {
        hamradio.publish('server/add', {name: this.state.addName, api: this.state.addAPI})
        this.setState({
          addName: '',
          addAPI: ''
        })
      })
      .catch(error => {
        this.setState({
          error: "Unable to reach API."
        })
        setTimeout(this.setState.bind(this), 3000, {error: null})
      })
  }

	render () {
		return (
      <div>
        <label>Add Server</label>
        <ul>
          {this.state.servers.map(server => <li key={server.api}>{server.name}: {server.api}</li>)}
        </ul>
        <div>
          <label>Nickname (optional)</label>
          <input type="text" name="server-add-name" value={this.state.addName} onChange={e => this.setState({addName:e.target.value})}/>
        </div>
        <div>
          <label>Server API address</label>
          <input type="text" name="server-add-api" value={this.state.addAPI} onChange={e => this.setState({addAPI:e.target.value})}/>
        </div>
        {this.state.error
          ? <label className="WarningLabel">Failed to add server: {this.state.error}</label>
          : <button className="btn btn-default Button" onClick={this.addServer}>Add Server</button>
        }
      </div>
		)
	}
}


export default ServerAdd;
