// components/CTracksComp/ServerAdd.js
import React from 'react'
import axios from 'axios'

import './ServerAdd.css'

import hamradio from 'hamradio'

import TextInput from './basicComponents/TextInput'
import LabelLayout from './basicComponents/LabelLayout'

class ServerAdd extends React.Component {
  constructor (props) {
    super(props)

    this.subscriptions = this.makeSubscriptions()

    this.state = {
      servers: [],
      addName: '',
      addAPI: '',
      error: null
    }
  }

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions = () => {
    return [
      hamradio.subscribe(
        'server/add',
        (name, server) => {
          this.setState({
            servers: this.state.servers.concat([server])
          })
        }
      )
    ]
  }

  addServer = () => {
    axios.get(`${this.state.addAPI}/tilesets`)
      .then(response => {
        hamradio.publish('server/add', {name: this.state.addName, api: this.state.addAPI})
        this.setState({
          addName: '',
          addAPI: ''
        })
      })
      .catch(error => {
        this.setState({
          error: "Unable to reach API."
        })
        setTimeout(this.setState.bind(this), 3000, {error: null})
      })
  }

	render () {
		return (
      <div>
        <label>Add Server</label>
        <ul>
          {this.state.servers.map(server => <li key={server.api}>{server.name}: {server.api}</li>)}
        </ul>
        <LabelLayout label="Nickname (optional)"
          inner=<TextInput value={this.state.addName} update={val => this.setState({addName:val})}/>
        />
        <LabelLayout label="Server API address"
          inner=<TextInput value={this.state.addAPI} update={val => this.setState({addAPI:val})}/>
        />
        {this.state.error
          ? <label className="WarningLabel">Failed to add server: {this.state.error}</label>
          : <button className="btn btn-default Button" onClick={this.addServer}>Add Server</button>
        }
      </div>
		)
	}
}


export default ServerAdd;
