// components/CTracksComp/TrackAdd.js
import React from 'react'
import axios from 'axios'
import hamradio from 'hamradio'

import './TrackAdd.css'

import tracks from './utils/tracks'

class TrackAdd extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      tilesets: [null],
      tileset: 0
    }
    this.subscriptions = this.makeSubscriptions()
  }

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions() {
    return [
      hamradio.subscribe(
        'server/updated',
        (name, server) => {
          this.getTilesets()
        }
      )
    ]
  }

  componentDidMount() {
    this.getTilesets()
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.servers !== this.props.servers ||
        prevProps.excludeTracks !== this.props.excludeTracks) {
      this.getTilesets()
    }
  }

  getTilesets = () => {
    let promises = this.props.servers.map(server => {
      return axios.get(`${server.api}/tilesets`)
        .then(response => {
          return {server, tilesets: response.data.results}
        })
        .catch(error => {
          console.log(error)
          return {server, tilesets:[]}
        })
    })
    Promise.all(promises).then(tilesetResults => {
      let tilesets = tilesetResults
        .map(({server, tilesets}) => {
          return tilesets
            .filter(tileset => !this.props.excludeTracks.includes(tileset.uuid) )
            .map(tileset => { return {server, tileset}})
        })
      this.setState({
        tilesets: [null].concat(...tilesets)
      })
    })
  }

  setStr = (set) => {
    console.log(set)
    if (!set) return ''
    let serverName = !set.server.name ? set.server.api : set.server.name
    let tilesetName = !set.tileset.name ? set.tileset.uuid : set.tileset.name
    return `${serverName}: ${tilesetName}`
  }

  done = () => {
    let tileset = this.state.tilesets[this.state.tileset]
    if (tileset) {
      hamradio.publish(
        'track/add',
        {
          track: tracks.track({
            server: tileset.server.api,
            tilesetUid: tileset.tileset.uuid,
            name: tileset.tileset.name,
            uid: `${tileset.tileset.uuid}-track`
          }),
          global: true,
          focus: true,
          order: this.props.order
        }
      )
    }
  }

	render () {
		return (
      <div>
        <button type="button" className="btn btn-default Button" onClick={this.done}>Add</button>
        <select value={this.state.tileset} onChange={(e) => this.setState({tileset: e.target.value})}>
          {this.state.tilesets.map((set, index) =>
            <option key={index} value={index}>
              {this.setStr(set)}
            </option>
          )}
        </select>
      </div>
    )
	}
}


export default TrackAdd;
