// components/CTracksComp/TracksMenu.js
import React from 'react'
import hamradio from 'hamradio'
import {CompactPicker} from 'react-color'
import _ from 'lodash'

import TrackAdd from './TrackAdd'
import LabelLayout from './basicComponents/LabelLayout'
import Selector from './basicComponents/Selector'
import IconButton from './basicComponents/IconButton'
import TextButton from './basicComponents/TextButton'

import viewconfig from './utils/viewconfig'
import tracks from './utils/tracks'

import './TracksMenu.css'


class TracksMenu extends React.Component {
  constructor (props) {
    super(props);

    this.subscriptions = this.makeSubscriptions()

    this.state = {
      editTrack: null,
      editRow: null
    }
	}

  componentWillUnmount() {
    this.subscriptions.forEach(sub => sub.unsubscribe())
  }

  makeSubscriptions() {
    return [
      hamradio.subscribe(
        'track/add',
        (name, row) => {
          if (!this.state.editTrack) {
            console.log(row)
            this.editTrack(row, Object.values(row.tilesets)[0])
          }
        }
      ),
      hamradio.subscribe(
        'track/modify',
        (name, row) => {
          if (this.state.editTrack && this.state.editTrack.uid in row.tilesets) {
            this.setState({editTrack: row.tilesets[this.state.editTrack.uid]})
          }
        }
      ),
      hamradio.subscribe(
        'track/remove',
        (name, track) => {
          if (this.state.editTrack && track.uid === this.state.editTrack.uid) {
            this.setState({editTrack: null, editRow: null})
          }
        }
      )
    ]
  }

  toggleVisibility = (row, global, focus) => {
    hamradio.publish('track/modify', {...row, global, focus})
  }

  editTrack = (row, track) => {
    this.setState({
      editTrack: _.cloneDeep(track),
      editRow: _.cloneDeep(row)
    })
  }

  setTrackValue = (key, value) => {
    const track = _.merge({}, this.state.editTrack, {[key]: value})
    const row = _.merge({}, this.state.editRow, {tilesets: {[this.state.editTrack.uid]: track}})
    hamradio.publish('track/modify', row)
  }

  setOptionValue = (option, value) => {
    const track = _.merge({}, this.state.editTrack, {options: {[option]: value}})
    const row = _.merge({}, this.state.editRow, {tilesets: {[this.state.editTrack.uid]: track}})
    hamradio.publish('track/modify', row)
  }

  setPrimaryColor = (color) => {
    let options = tracks.options[this.state.editTrack.type]({}, color)
    const track = _.merge({}, this.state.editTrack, {options})
    const row = _.merge({}, this.state.editRow, {tilesets: {[this.state.editTrack.uid]: track}})
    hamradio.publish('track/modify', row)
  }

  changeType = (type) => {
    let primaryColor = tracks.getPrimaryColor(this.state.editTrack)
    let options = tracks.options[type]({}, primaryColor)
    const track = _.merge({}, this.state.editTrack, {type, options})
    const row = _.merge({}, this.state.editRow, {tilesets: {[this.state.editTrack.uid]: track}})
    hamradio.publish('track/remove', {row: this.state.editRow.row, uid: this.state.editTrack.uid})
    hamradio.publish('track/add', row)
  }

  trackOverview = () => {
    return (
      <div>
  			<label>Tracks</label>
        {viewconfig.sortTracks(Object.values(this.props.tracks)).map(row => (
          <div key={row.row}>
            <LabelLayout label={`Row ${row.row}`}
              inner=<div>
                <TextButton
                  onClick={() => this.toggleVisibility(row, !row.global, row.focus)}
                  text="Global"
                  backgroundColor={row.global ? 'black' : 'white'}
                  color={row.global ? 'white' : 'grey'}
                />
                <TextButton
                  onClick={() => this.toggleVisibility(row, row.global, !row.focus)}
                  text="Focus"
                  backgroundColor={row.focus ? 'black' : 'white'}
                  color={row.focus ? 'white' : 'grey'}
                />
              </div>
            />
            <ul>
              {Object.values(row.tilesets).map(tileset => (
                <li key={tileset.uid}>
                  <IconButton icon="eye" backgroundColor={tracks.getPrimaryColor(tileset)} color="white" onClick={() => this.editTrack(row, tileset)}/>
                  <label>{tileset.name}</label>
                </li>
              ))}
              <li>
                <TrackAdd
                  servers={this.props.servers}
                  row={row.row}
                />
              </li>
            </ul>
          </div>
        ))}
        <TrackAdd
          servers={this.props.servers}
          row={Math.max(0, ...Object.values(this.props.tracks).map(row => row.row))+1}
        />
      </div>
		)
  }

  trackTypeEdit = (type) => {
    return (
      <div>
        <div className="ColorPicker">
          <CompactPicker color={tracks.getPrimaryColor(this.state.editTrack)} onChange={color => this.setPrimaryColor(color.hex)}/>
        </div>
        {type === 'range'
          ? <div>
              <LabelLayout label="Resolution"
                inner=<input type="number" name="resolution" min="1" max="10"
                  value={this.state.editTrack.options.resolution || 1} onChange={e => this.setOptionValue('resolution', e.target.value)}
                />
              />
              <LabelLayout label="Display as"
                inner=<div>
                  <Selector
                    value={{value: this.state.editTrack.options.mode, label: this.state.editTrack.options.mode}}
                    onChange={option => this.setOptionValue('mode', option.value)}
                    options={[
                      {value: 'minMax', label: 'minMax'},
                      {value: 'whisker', label: 'whisker'}
                    ]}
                  />
                </div>
              />
            </div>
          : null
        }
      </div>
    )
  }

  trackEdit = () => {
    return (
      <div>
  			<label>Higlass - Edit Track</label>
        <div>
          <ul>
            <li>{this.state.editTrack.server}</li>
            <li>{this.state.editTrack.name}</li>
          </ul>
          <LabelLayout label="Track Type"
            inner=<div>
              <Selector
                value={{value: this.state.editTrack.type, label: this.state.editTrack.type}}
                onChange={option => this.changeType(option.value)}
                options={tracks.trackDataTypes[this.state.editTrack.datatype]
                  .concat(this.state.editTrack.filetype === 'bigwig' ? ['range'] : [])
                  .map(type => ({value: type, label: type}))}
              />
            </div>
          />
          {this.trackTypeEdit(this.state.editTrack.type)}
          <br/>
          <TextButton onClick={() => this.editTrack(null, null)} text="Done" />
        </div>
      </div>
    )
  }

	render () {
		return !this.state.editTrack ? this.trackOverview() : this.trackEdit()
	}
}


export default TracksMenu;
