// components/CTracksComp/basicComponents/Selector.js
import React from 'react'
import Select from 'react-select'

const style = {
  option: (provided, state) => ({
    ...provided,
    marginTop: "0px",
    marginBottom: "0px",
    padding: "0px",
    minHeight: "20px"
  }),
  noOptionsMessage: (provided, state) => ({
    ...provided,
    marginTop: "0px",
    marginBottom: "0px",
    padding: "0px",
    minHeight: "20px"
  }),
  input: (provided, state) => ({
    ...provided,
    width: "100%",
    padding: "2px",
    margin: "0px",
    minHeight: "20px"
  }),
  indicatorsContainer: (provided, state) => ({
    ...provided,
    padding: "0px",
    margin: "0px",
    minHeight: "20px"
  }),
  dropdownIndicator: (provided, state) => ({
    ...provided,
    padding: "0px",
    margin: "0px",
    minHeight: "20px"
  }),
  container: (provided, state) => ({
    ...provided,
    width: "100%",
    padding: "0px",
    margin: "0px",
    minWidth: "150px"
  }),
  control: (provided, state) => ({
    ...provided,
    padding: "0px",
    margin: "0px",
    minHeight: "20px"
  }),
  valueContainer: (provided, state) => ({
    ...provided,
    padding: "0px",
    margin: "0px",
    minHeight: "20px"
  })
}

export default function Selector (props) {
  return <Select
    value={props.value}
    options={props.options}
    onChange={props.onChange}
    styles={style}
  />
}
